src/request_tracing.cpp

```diff
@@ -107,51 +158,73 @@ RequestTracing::RequestTracing(ngx_http_request_t *request,
   // and so no `RequestTracing` objects are ever instantiated.
   assert(main_conf_);
 
-  auto tracer = ot::Tracer::Global();
+  auto *tracer = global_tracer();
   if (!tracer) throw std::runtime_error{"no global tracer set"};
 
-  std::unique_ptr<ot::SpanContext> extracted_context = nullptr;
-  if (parent_span_context == nullptr && loc_conf_->trust_incoming_span) {
-    extracted_context = extract_span_context(*tracer, request_);
-    parent_span_context = extracted_context.get();
-  }
-
   ngx_log_debug1(NGX_LOG_DEBUG_HTTP, request_->connection->log, 0,
                  "starting Datadog request span for %p", request_);
-  request_span_ = tracer->StartSpan(
-      get_request_operation_name(request_, core_loc_conf_, loc_conf_),
-      {ot::ChildOf(parent_span_context),
-       ot::StartTimestamp{to_system_timestamp(request->start_sec, request->start_msec)}});
-  if (!request_span_) throw std::runtime_error{"tracer->StartSpan failed"};
+
+  dd::SpanConfig config;
+  auto start_timestamp = to_system_timestamp(request->start_sec, request->start_msec);
+  config.start = estimate_past_time_point(start_timestamp);
+  config.name = get_request_operation_name(request_, core_loc_conf_, loc_conf_);
+
+  if (!parent && loc_conf_->trust_incoming_span) {
+    NgxHeaderReader reader{request};
+    auto maybe_span = tracer->extract_or_create_span(reader, config);
+    if (auto *error = maybe_span.if_error()) {
+      ngx_log_error(NGX_LOG_ERR, request->connection->log, 0,
+                    "failed to extract a Datadog span request %p: [error code %d]: %s", request,
+                    error->code, error->message.c_str());
+    } else {
+      request_span_.emplace(std::move(*maybe_span));
+    }
+  }
+
+  if (!request_span_) {
+    if (parent) {
+      request_span_.emplace(parent->create_child(config));
+    } else {
+      request_span_.emplace(tracer->create_span(config));
+    }
+  }
 
   if (loc_conf_->enable_locations) {
     ngx_log_debug3(NGX_LOG_DEBUG_HTTP, request_->connection->log, 0,
                    "starting Datadog location span for \"%V\"(%p) in request %p",
                    &core_loc_conf->name, loc_conf_, request_);
-    span_ = tracer->StartSpan(get_loc_operation_name(request_, core_loc_conf_, loc_conf_),
-                              {ot::ChildOf(&request_span_->context())});
-    if (!span_) throw std::runtime_error{"tracer->StartSpan failed"};
+    dd::SpanConfig config;
+    config.name = get_loc_operation_name(request_, core_loc_conf_, loc_conf_);
+    span_.emplace(request_span_->create_child(config));
```

Comment:
It might be achievable, but it deserves deeper consideration (preferably at a later time). Ideally locations would be nested within each other (or perhaps not), whereas now they're all siblings of each other.

The default and most desired behavior is one span per request. In the future we can revisit the purpose of location spans, dedicated "upstream" spans, and possibly spans corresponding to the work of other modules.